### ISSUE TITLE
Avoid theorems with conjunctions in the consequent

### ISSUE THREAD

**Issue by avekens** [metamath/set.mm#1325]
In PR #1320 , Benoît gave the following remark:

> @avekens , @nmegill : Wouldn't it be more flexible to state ~2fvidf1o as two theorems? I think it should be most often the case when the (consequent of the conclusion of the) theorem is a conjunction ?
> Edit: ~m2cpminv is in this case.

I fully agree with Benoît, although there is one reason for keeping such theorems together: if the two parts share a big part of the proof, i.e. separating the theorem in two (or more) theorems would result in duplicated (parts of the) proofs. 

Nevertheless, I would like to add the convention that theorems with conjunctions in the consequent should be avoided (starting from Part 2 of set.mm). What do others think about such a convention?

In the special case of ~ 2fvidf1o , I decided to provide such a combined theorem because the main theorem I used in the proof is alread such a combined theorem: ~ fcof1o . So following the new convention, ~ fcof1o would have to be split into theorems already. I'll make a proposal for this in my next PR. If ~ fcof1o was split, the proof of ~ pmtrff1o , for example, could be shortened (by 1 step, ~ simpld ), because it only requires the left hand side of the conclusion in ~ fcof1o .

**Comment by digama0** [comment]
I don't support this convention. As you point out, it is very often the case that two facts are proven in tandem, for example an existential uniqueness. In that case of course we have a definition, but more complex situations often arise and I would not want to impose a constraint on what kinds of lemmas to have - anything that makes the proof simpler should be supported.

That said, this is specifically for lemmas and large proofs. If the theorem gets any significant amount of use, I would consider it common courtesy to weaken the theorem to its conjuncts, as *new* theorems derived from the original conjunction theorem. This approach allows you to separate concerns between the lemma, where ergonomics of the proof itself are primary, and the theorem, where the ergonomics of use are more important.

If a theorem is just two separate theorems, with separate proofs, stapled together, then of course there is no reason to have them be one theorem, as it is undoubtedly less convenient.

One other example where conjunctions come up is in algorithmic proofs; the invariant in an algorithm may well be a conjunction of some sort. ~ prmlem0 isn't a perfect example, because it's not a conjunction, but it does prove a very specific statement without any of the usual trappings (it has an unnecessary set variable `x` in the conclusion, for example) which is written that way so that it can have the same form as its first hypothesis.

**Comment by digama0** [comment]
In the specific case of ~ fcof1o , I think it would be reasonable to prove the left conjunct version of that theorem. The conjunction is deliberate, because you get both facts from all the assumptions and less than that is information-lossy, but often we only care about the left conjunct, because we don't really care about the function G except insofar as it shows that F is bijective.

This is just an ergonomics question, so it depends on how often the theorem gets used. Assuming that it was really popular, I would prefer to see it in deduction form with just the left part, proven from fcof1o.

**Comment by benjub** [comment]
I agree with your arguments.  It should be more a rule of thumb than a convention.  Probably, it might often be advantageous to extract the common part of the proofs in a lemma
> lem $p |- ( A -> D )

and then prove
> thm1 $p |- ( A -> B )
> thm2 $p |- ( A -> C )

from the lemma, rather than proving directly
> thm  $p |- ( A -> ( B /\ C ) )

and it is then more flexible to use thm1/2 than thm.  But I agree that there are cases were it is preferable to have the latter (although it is proved in one step from thm1/2), especially when B and C will undergo common transformations.

And for all of this, the deduction versions are more flexible indeed, when there are hypotheses (that is, the conclusion of thm1 would be thm1 $p |- ( ph -> ( A -> B ) ) etc.).

**Comment by nmegill** [comment]
Like @digama0 I also don't think it would be a good idea to do this generally.  For one thing, we should make at least some attempt to prevent set.mm from becoming larger than it has to be, and little things like this can add up.  Splitting the theorem saves a single proof step (simpld or simprd) in a proof referencing it, and it would take many uses to make up for the overhead of stating the theorem 3 times instead of once.

I also question whether it would necessarily be better for the user of the theorem.  For one thing, there would now be two theorems that have to be remembered instead of one.  A single theorem informs the user that, given the hypotheses and antecedents, both consequents follow from them, which I think is useful information.  In case both consequents are needed in a proof, the user doesn't have to check a second time that all assumptions are satisfied.  Having to apply simpld or simprd just isn't that much of a nuisance or overhead and can usually be figured out automatically by the proof assistant.

Historically we've tended to break up theorems this way only when there have been enough uses to justify it.  For example, f1ofo was added in 2004, and before that, dff1o3 (1998) or df-f1o (1994) was used instead.  At the time, looking at an old set.mm, f1ofo resulted in slightly shortening 12 proofs.  Retrofitting proofs to use f1ofo was trivial by applying 'minimize f1ofo' to all proofs using dff1o3 or df-f1o.  (Today f1ofo is used 190 times so it clearly pays off.)

Right now, 2fvidf1o is used only once (in m2cpminv), and in fact this single use requires both consequents.  So splitting 2fvidf1o would only make the proof longer, not shorter, in addition to the overhead of splitting it.  And at this point there would be no purpose in splitting m2cpminv since it isn't even used.

Splitting a theorem like this is trivial whenever it becomes advantageous to do so, and I don't see a reason to do it in advance until then.

**Comment by benjub** [comment]
You made a good point: if one has theorems "if hypotheses H1 to H9 are satisfied, then A, and also, if hypotheses H1 to H9 are satisfied, then B", it makes sense to group them. Also, I believe you that it could be generally more efficient to group such theorems in set.mm.

My concern was to have each metamath-theorem correspond to a single mathematical fact (a vague notion, I confess; in the first example above, the conjunction might be a single fact).  In particular, I was prompted by a theorem I had seen earlier: ~nthruc , which to me expresses four different facts (and even, each strict inclusion is actually two facts, the inclusion being the structurally important fact -- not to say that the inequality is not important!).

**Comment by nmegill** [comment]
@benjub In the case of nthruc, I purposely put them in one theorem so the reader can grasp visually and immediately the proper subset chain formed by the 5 sets of numbers.  I think this point would be harder to see or even lost if we broke it up into 4 (or 8) separate theorems scattered throughout set.mm.  So far, nthruc hasn't had any uses (and I didn't expect it to), but it would be trivial to extract any particular piece if needed.  If the piece is used frequently enough, it would become its own theorem in the same way I described for f1ofo.  (The non-proper subset pieces of nthruc are already their own theorems of course.)

Gleason's Fundamentals of Abstract Analysis also has this chain of proper inclusions on one line on p. 132, to make the same point I assume.  I should probably add that reference to the description.

**Comment by avekens** [comment]
OK, you convinced me that it makes no sense to avoid theorems with conjunctions in the consequent in general. 

For ~ fcof1o , however, I followed @digama0`s  hint to split it into 2 theorems and provide them in deduction form: ~ fcof1oinvd and ~ fcof1od (see PR #1327). The proof of ~ fcof1o could be split into two almost independent parts, resulting in the proofs of the new theorems. By this, set.mm is not essentially enlarged (except by providing more comments...). The proof for ~ fcof1o , which I kept, could be shortened to be a combination of the 2 new theorems. I think ~ fcof1o could be replaced by ~ fcof1od in the proofs of some other theorems. 

Analogously, I split ~ 2fvidf1o into 2 theorems ~ 2fvidinvd and ~ 2fvidf1od (in deduction form). Here, I did not keep the combined, former theorem ~ 2fvidf1o, but combined the 2 new theorems to prove  ~ m2cpminv , which is provided two show that ( N cPolyMatToMat R ) is the inverse function of  ( N matToPolyMat R ) , analogously to the purpose of ~ nthruc .

I hope this is acceptable for everyone.

**Comment by avekens** [comment]
By the way, one usage of the discouraged "4syl" was removed by my modifications.

**Comment by nmegill** [comment]
I just noticed that very strangely, I didn't receive an email for my message at https://github.com/metamath/set.mm/issues/1325#issuecomment-565832050 .  Did anyone else NOT receive it?

> By the way, one usage of the discouraged "4syl" was removed by my modifications.

4syl is discouraged only because it sometimes causes an excessive slowdown of 'minimize_with' (see its description).  Otherwise people are free to use it (or not).  Incidentally the MM-PA command 'expand 4syl' will remove 4syl from a proof, and 'minimize 4syl/override' will put it back in.

**Comment by avekens** [comment]
@nmegill I got an email  for your message at [#1325 (comment)  ](https://github.com/metamath/set.mm/issues/1325#issuecomment-565832050) ...

I think by PR #1327 this issue is done and can be closed.
